include/envoy/upstream/cluster_manager.h

```diff
@@ -350,6 +358,26 @@ class ClusterManagerFactory {
    * Returns the secret manager.
    */
   virtual Secret::SecretManager& secretManager() PURE;
+
+  /**
+   * Returns a struct with all the Stats::StatName objects needed by
+   * ClusterManager. This helps factor out some relatively heavy name
+   * construction which occur when there is a large CDS update during operation,
+   * relative to recreating all stats from strings on-the-fly.
+   *
+   * @return the stat names.
+   */
+  virtual const ClusterManagerStatNames& clusterManagerStatNames() const PURE;
+
+  /**
+   * Returns a struct with all the Stats::StatName objects needed by the
+   * router. This helps factor out some relatively heavy name construction which
+   * occurs when there is a large CDS update during operation, relative to
+   * recreating all stats from strings on-the-fly.
+   *
+   * @return the stat names.
+   */
+  virtual const Router::RouterStatNames& routerStatNames() const PURE;
```

Comment:
we hang the routerStatNames off the clusterManagerFactory because we happen to have that available in the place where we construct the router's FilterConfig. It's arguably not too clean. I could also just add a RouterContext and plumb that through the chain of constructors.

Reply:
From a cleanliness/tech debt perspective, we are trying to remove protocol specific stuff from cluster manager (@alyssawilk just started on some of this in terms of moving hard coded HTTP logic out to extensions).

It seems to me the best way to do this would be create a singleton somewhere where the router config is created? Then it would just be created once and lookup?

Reply:
I  think I have a better place to put it: the Http::Context. I thought of that because of your comment about 'hard coded HTTP logic". Does that seem reasonable?

And I have an Http::Context available to me where I needed the router stat names without excess plumbing.

Reply:
This is very router filter specific though. Can't it just be a singleton provided by the router filter factory?

Reply:
Per slack, this is not the router filter, it was in source/common/router.

I earlier had this PR in a state where I had put the names needed for Router::FilterStats into Http::Context, and that was pretty straightforward, but wasn't great for separation. Now I have a Router::Context. Plumbing that through was trivial but involved a lot of file edits due to parallel universes for mocks, validation servers, integration test servers, and whatnot. 

LMK what you think; I'm OK trying this out as a prod singleton too, but I think it would involve a few implicit dependencies that are now explicit (if verbose). For example in every test method we'd need to re-init the singleton with the test's symbol table.

Reply:
Yeah sorry I had forgotten the AsyncClient angle to this. Given that this is basically part of the core I'm fine putting it wherever you think is easiest.

Reply:
I think it's a bit cleaner to have the router stats in their won context, though it expanded the number of files in the PR a bit.